### PR TITLE
[devscripts] refactoring role to align with layout and network attachment definitions

### PR DESCRIPTION
**Pull request by psathyan**
The refactoring was done to ensure `devscripts` role performs only tasks related to deploying OpenShift platform. At a high level, the following changes are done

- Aligning to use network attachment definitions
- Layout is driven by `cifmw_libvirt_manager_configuration`
- Supports more than 4 NIC attachments
- Supports user provided external network configuration
- Support for external static ip address

The following support has been removed
- Adding of baremetal host
- Using the default network as external network
- configuring bridge VLANs
- compute node information (as it is driven by layout)
- Error message when deprecated variables for `ci_token` and `pull_secret` is used

*Workflow*
```
    100_pre
        110 -> check for existing OpenShift cluster (be it running or existing golden definition)
        120 -> Check the passed values of ci_token & pull_secret
        130 -> Prepare the target host
            131 -> Specific to downstream... extend root_lvm disk . Default set to 70G
            132 -> Network configuration (host, dns & removal of default-virt)
            133 -> Create the required folders and clone the repo
            134 -> Patch dev-scripts repo (proxy & baremetal ip address modification to vm_setup_vars)
            135 -> Copy the secrets
            136 -> Customization of OpenShift deployment
            137 -> Trigger when static IP address is required
            138 -> Copy

   deploy
        - Trigger dev-scripts make

    300_post
        - Gather and set the cluster access information
        310 -> Prepare the environment for disk overlay
        320 -> Prepares the test environment layout
        - Deploy the layout
        330 -> Wait till the OpenShift cluster is accessible.
```

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [x] Appropriate documentation exists and/or is up-to-date:
  - [x] README in the role
  - [x] Content of the docs/source is reflecting the changes

*Test Scenarios*
- [x] Test with `deploy-edpm.yml`
- [x] Test cleanup with `deepscrub`
- [x] Test without `deepscrub`
- [x] Test with `reproducer.yml`
- [x] Test with `reproducer-clean.yml`

*Reproducer run*
```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
titanamd126                : ok=533  changed=165  unreachable=0    failed=0    skipped=178  rescued=1    ignored=0   

Wednesday 14 February 2024  08:47:49 -0500 (0:00:00.139)       1:07:06.926 **** 
==============================================================================
[root@titanamd126 ~]# virsh list --all
 Id   Name                 State
------------------------------------
 8    cifmw-compute-0      running
 9    cifmw-compute-1      running
 10   cifmw-compute-2      running
 11   cifmw-controller-0   running
 12   cifmw-ocp-0          running
 13   cifmw-ocp-1          running
 14   cifmw-ocp-2          running

[root@titanamd126 ~]# 
[root@titanamd126 ~]# 
[root@titanamd126 ~]# su - ciuser
Last login: Wed Feb 14 14:42:20 IST 2024 from 10.0.100.219 on pts/0
[ciuser@titanamd126 ~]$ ssh controller-0
Warning: Permanently added '192.168.111.50' (ED25519) to the list of known hosts.
Register this system with Red Hat Insights: insights-client --register
Create an account or view all your systems at https://red.ht/insights-dashboard
[zuul@controller-0 ~]$ oc get nodes
NAME       STATUS   ROLES                         AGE   VERSION
master-0   Ready    control-plane,master,worker   46m   v1.27.10+28ed2d7
master-1   Ready    control-plane,master,worker   46m   v1.27.10+28ed2d7
master-2   Ready    control-plane,master,worker   46m   v1.27.10+28ed2d7
[zuul@controller-0 ~]$ 
```

### CODE REVIEW

**Comment by cjeanner** [comment]
/hold

We have to double-check with latest #1151 which fixes how the overlays are used with OCP cluster. Maybe re-basing and doing some new tests?

**Comment by psathyan** [comment]
Testing results after cherry picking #1151 

*First run*
```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
titanamd126                : ok=372  changed=137  unreachable=0    failed=0    skipped=166  rescued=1    ignored=0   

Friday 16 February 2024  00:14:38 -0500 (0:00:00.128)       1:04:09.910 ******* 
=============================================================================== 
devscripts : Deploy OpenShift cluster ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2665.13s
reproducer : Install internal CA ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 175.39s
reproducer : Install some tools ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 118.21s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.41s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.02s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 50.64s
reproducer : Bootstrap environment on controller-0 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 32.13s
networking_mapper : Ensure that networking and hostname facts are in place ----------------------------------------------------------------------------------------------------------------------------------------------------------- 30.92s
libvirt_manager : Grab IPs for nodes type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 28.62s
reproducer : Install collections on controller-0 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 25.14s
libvirt_manager : Grab IPs for nodes type compute ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 20.30s
libvirt_manager : Grab IPs for nodes type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.93s
libvirt_manager : Wait for SSH on VMs type ocp --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.91s
libvirt_manager : Prepare disk image with SSH and growing volume --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.31s
libvirt_manager : Wait for SSH on VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.86s
libvirt_manager : Configure VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.05s
reproducer : Wait for OCP nodes to be ready ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 16.66s
reproducer : Inject ProxyJump configuration for remote hypervisor VMs ---------------------------------------------------------------------------------------------------------------------------------------------------------------- 15.47s
reproducer : Install ansible dependencies -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 14.88s
devscripts : Dump XMLs in a file for all nodes part of OpenShift platform ------------------------------------------------------------------------------------------------------------------------------------------------------------ 12.11s
```

File: roles/devscripts/files/proxy.patch
Comment (cjeanner) [inline roles/devscripts/files/proxy.patch:1]: I think this one isn't needed anymore?

File: roles/devscripts/tasks/110_check_ocp.yml [hunk 1]
```diff
@@ -14,27 +14,25 @@
...
-- name: Using existing deployment layer for creating OCP cluster.
+- name: Verify the node definition match with expected count.
   when: >-
     (
       cifmw_devscripts_config.num_masters | int +
       cifmw_devscripts_config.num_workers | int
     ) == overlay_results.matched
   ansible.builtin.set_fact:
-    cifmw_devscripts_use_base_layer: true
+    cifmw_devscripts_ocp_exists: true
```
Comment (cjeanner) [inline roles/devscripts/tasks/110_check_ocp.yml:32]: you can use the `when` condition as the actual value here, allowing to remove the `|default(false)` maybe?

File: roles/devscripts/tasks/110_check_ocp.yml [hunk 2]
```diff
@@ -64,22 +62,11 @@
         username: "{{ cifmw_openshift_user }}"
         validate_certs: false
       register: _login_result
+      ignore_errors: true
```
Comment (cjeanner) [inline roles/devscripts/tasks/110_check_ocp.yml:65]: iirc ansible-lint will complain. Maybe `failed_when: false` is better.

File: roles/devscripts/tasks/120_verify_params.yml
Comment (cjeanner) [inline roles/devscripts/tasks/120_verify_params.yml:1]: I think most of those checks are already done in the `manage_secrets` - if any is missing, they should be nudged in that role instead.

File: roles/devscripts/tasks/130_prep_host.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Ensure the required packages for executing dev-scripts exists.
+  tags:
+    - bootstrap
+    - packages
+  become: true
+  ansible.builtin.package:
```
Comment (cjeanner) [inline roles/devscripts/tasks/130_prep_host.yml:22]: note that there are some packages coming from EPEL. In the reproducer, I could get around it with this:
https://github.com/openstack-k8s-operators/ci-framework/blob/main/roles/reproducer/tasks/devscripts_layout.yml#L2-L11

File: scenarios/reproducers/validated-architecture-1.yml [hunk 4]
```diff
@@ -79,7 +79,7 @@ cifmw_libvirt_manager_configuration:
       admin_user: core
       image_local_dir: "/home/dev-scripts/pool"
       disk_file_name: "ocp_master"
-      disksize: "105"
+      disksize: "100"
```
Comment (cjeanner) [inline scenarios/reproducers/validated-architecture-1.yml:82]: bumping it slightly above the actual size ensure we're not truncating it.

File: scenarios/reproducers/validated-architecture-1.yml [hunk 5]
```diff
@@ -90,14 +90,14 @@ cifmw_libvirt_manager_configuration:
 
 ## devscript support for OCP deploy
 cifmw_use_devscripts: true
-cifmw_devscripts_ocp_version: '4.13.13'
+cifmw_devscripts_ocp_version: '4.14.12'
```
Comment (cjeanner) [inline scenarios/reproducers/validated-architecture-1.yml:93]: I was wondering if we shouldn't rely on the default value set in the role?

File: roles/devscripts/tasks/110_check_ocp.yml (hunk 1 above)
Comment (psathyan) [reply roles/devscripts/tasks/110_check_ocp.yml:37]: Fixed in the next commit.

File: roles/devscripts/tasks/120_verify_params.yml
Comment (psathyan) [reply roles/devscripts/tasks/120_verify_params.yml:1]: Removed the conditions that check for file or content. Left behind the conditions that ensures we have pull_secret and ci_token passed.

File: cleanup-edpm.yml
[the remark this answers is not in the document]
Comment (cjeanner) [reply cleanup-edpm.yml:1]: right, indeed.

**Comment by cjeanner** [comment]
/hold cancel
Still a big piece to review... @cescgina do you think you may craft a reproducer job based on dev-scripts to have a quick check?

**Comment by cescgina** [comment]
> /hold cancel Still a big piece to review... @cescgina do you think you may craft a reproducer job based on dev-scripts to have a quick check?

sure, I have a machine idle while I fix other issues

File: roles/devscripts/tasks/135_patch_src.yml [hunk 6]
```diff
@@ -49,14 +31,18 @@
     _vm_setup_vars: "{{ vm_setup_data.content | b64decode | from_yaml }}"
 
 # Note: Override When external network is being reused.
-- name: Verify dev-scripts vm_setup_vars to reflects external network address
+- name: Ensure dev-scripts vm_setup_vars reflects external network address
```
Comment (pablintino) [inline roles/devscripts/tasks/135_patch_src.yml:34]: You can perfectly squash this task with the next one resulting in a simpler approach, without the useless `_modify_vm_setup` variable and the need to run two tasks (even three, cause the previous set_fact is useless too)

File: roles/devscripts/tasks/135_patch_src.yml
Comment (pablintino) [inline roles/devscripts/tasks/135_patch_src.yml:1]: Based on some errors chatted about in slack, it seems that devscripts has some checks, that sometimes fails, about ensuring its repo is untouched. Do you think its worthy to commit these changes?

File: cleanup-edpm.yml [hunk 7]
```diff
@@ -5,12 +5,18 @@
 # Note that this playbook can be called as a standalone tool.
 
 - name: Cleanup EDPM
-  hosts: localhost
+  hosts: "{{ cifmw_target_host | default('localhost') }}"
   gather_facts: true
+
   tasks:
+    # TODO: clean crc, libvirt when the roles are ready
```
[the remark this answers is not in the document]
Comment (psathyan) [reply cleanup-edpm.yml]: Sure, will remove this part

File: roles/devscripts/tasks/135_patch_src.yml
Comment (psathyan) [reply roles/devscripts/tasks/135_patch_src.yml:1]: I have submitted an upstream patch... sadly, there is movement. Let me try to rebase the upstream patch and get it merged.

Comment (psathyan) [reply roles/devscripts/tasks/135_patch_src.yml:1]: https://github.com/openshift-metal3/dev-scripts/pull/1607

File: roles/devscripts/tasks/135_patch_src.yml (hunk 6 above)
Comment (psathyan) [reply roles/devscripts/tasks/135_patch_src.yml:30]: Thanks.
Yes, that make sense.

**Pull request merged**